File: src-tauri/src/social/attach.rs
//! Pure helpers for DM attachments (ROADMAP T9c). Deterministic and IO-free so
//! they're exhaustively unit-testable; the network glue (presign → PUT bytes →
//! presigned GET) lives in `commands.rs`. The server caps an attachment at
//! 25 MiB (`ATTACHMENT_MAX_BYTES`), so we reject early client-side to avoid a
//! wasted upload round-trip.

/// Largest attachment the server will presign (mirrors `ATTACHMENT_MAX_BYTES`
/// in the server's `social_api.rs`: 25 MiB).
pub const MAX_ATTACHMENT_BYTES: u64 = 25 * 1024 * 1024;

/// The final path component of a user-picked file path, handling both `/` and
/// `\` separators (a Windows path may arrive on a Linux build and vice-versa).
/// Returns an empty string if the path ends in a separator.
pub fn basename(path: &str) -> String {
    path.rsplit(['/', '\\']).next().unwrap_or("").to_string()
}

/// Best-effort MIME type from a filename's extension, for the presign request's
/// `contentType` (and the `Content-Type` header on the PUT). Unknown/extension-
/// less files fall back to `application/octet-stream`, which the server accepts.
pub fn guess_content_type(name: &str) -> &'static str {
    let ext = name.rsplit('.').next().unwrap_or("").to_ascii_lowercase();
    match ext.as_str() {
        "png" => "image/png",
        "jpg" | "jpeg" => "image/jpeg",
        "gif" => "image/gif",
        "webp" => "image/webp",
        "bmp" => "image/bmp",
        "svg" => "image/svg+xml",
        "pdf" => "application/pdf",
        "txt" | "log" => "text/plain",
        "json" => "application/json",
        "zip" => "application/zip",
        "mp4" => "video/mp4",
        "webm" => "video/webm",
        "mp3" => "audio/mpeg",
        "ogg" => "audio/ogg",
        "wav" => "audio/wav",
        _ => "application/octet-stream",
    }
}

/// Whether `size` (bytes) is an acceptable, non-empty attachment. An empty file
/// is rejected because the server's presign requires `size > 0`.
pub fn is_acceptable_size(size: u64) -> bool {
    size > 0 && size <= MAX_ATTACHMENT_BYTES
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn basename_strips_both_separators() {
        assert_eq!(basename("/home/me/shot.png"), "shot.png");
        assert_eq!(basename(r"C:\Users\me\My Pics\shot.png"), "shot.png");
        assert_eq!(basename("plain.txt"), "plain.txt");
        assert_eq!(basename("/trailing/"), "");
    }

    #[test]
    fn content_type_known_and_unknown() {
        assert_eq!(guess_content_type("a.PNG"), "image/png");
        assert_eq!(guess_content_type("photo.jpeg"), "image/jpeg");
        assert_eq!(guess_content_type("clip.mp4"), "video/mp4");
        assert_eq!(guess_content_type("save.dat"), "application/octet-stream");
        assert_eq!(guess_content_type("noext"), "application/octet-stream");
    }

    #[test]
    fn size_bounds() {
        assert!(!is_acceptable_size(0));
        assert!(is_acceptable_size(1));
        assert!(is_acceptable_size(MAX_ATTACHMENT_BYTES));
        assert!(!is_acceptable_size(MAX_ATTACHMENT_BYTES + 1));
    }
}

File: mobile/src/core/attach.ts
// Chat attachments, phone side. A deliberate mirror of the desktop client's
// src-tauri/src/social/attach.rs: the same limit, the same MIME table, the same
// rejections. Two implementations of one contract is already one too many, so
// where they must both exist they should at least give the same answers.
//
// Upload is presign -> PUT the bytes to the returned URL -> send a chat frame
// carrying the attachment id. The presign is the only authenticated call; the
// upload URL carries its own signature, so the session token must never be
// attached to it.

/** Largest attachment the server will presign — 25 MiB, matching the server's
 *  ATTACHMENT_MAX_BYTES and the desktop's MAX_ATTACHMENT_BYTES. Checked here so
 *  a photo that is too big fails instantly instead of after a long upload on a
 *  phone connection. */
export const MAX_ATTACHMENT_BYTES = 25 * 1024 * 1024;

/** The last component of a path or content URI, with both separators handled. */
export function basename(path: string): string {
  const cut = path.split(/[/\\]/);
  return cut[cut.length - 1] ?? "";
}

/** Best-effort MIME type from the extension. Unknown types fall back to
 *  application/octet-stream, which the server accepts. */
export function guessContentType(name: string): string {
  const ext = (name.split(".").pop() ?? "").toLowerCase();
  switch (ext) {
    case "png":
      return "image/png";
    case "jpg":
    case "jpeg":
      return "image/jpeg";
    case "gif":
      return "image/gif";
    case "webp":
      return "image/webp";
    case "bmp":
      return "image/bmp";
    case "svg":
      return "image/svg+xml";
    case "pdf":
      return "application/pdf";
    case "txt":
    case "log":
      return "text/plain";
    case "json":
      return "application/json";
    case "zip":
      return "application/zip";
    case "mp4":
      return "video/mp4";
    case "webm":
      return "video/webm";
    case "mp3":
      return "audio/mpeg";
    case "ogg":
      return "audio/ogg";
    case "wav":
      return "audio/wav";
    default:
      return "application/octet-stream";
  }
}

/** An empty file is rejected because the server's presign requires size > 0. */
export function isAcceptableSize(size: number): boolean {
  return Number.isFinite(size) && size > 0 && size <= MAX_ATTACHMENT_BYTES;
}

/** Why a picked file cannot be sent, phrased for the user. null = go ahead. */
export function attachmentBlocker(name: string, size: number): string | null {
  if (!basename(name)) return "That file has no name.";
  if (size <= 0) return "That file is empty.";
  if (size > MAX_ATTACHMENT_BYTES) return `Attachments are limited to ${formatSize(MAX_ATTACHMENT_BYTES)}.`;
  return null;
}

/** The presign request body. camelCase to match the desktop's PresignReq, which
 *  is what the server's handler actually reads. */
export interface PresignRequest {
  filename: string;
  contentType: string;
  size: number;
}

export function presignRequest(name: string, size: number): PresignRequest {
  const filename = basename(name);
  return { filename, contentType: guessContentType(filename), size };
}

export interface Presigned {
  attachmentId: number;
  uploadUrl: string;
}

/** Narrow the presign response. Both fields are required: an id with no URL
 *  would leave us sending a chat frame pointing at bytes that were never
 *  uploaded, which is worse than failing here. */
export function parsePresign(value: unknown): Presigned | null {
  if (!value || typeof value !== "object") return null;
  const v = value as Record<string, unknown>;
  const attachmentId = typeof v.attachmentId === "number" ? v.attachmentId : 0;
  const uploadUrl = typeof v.uploadUrl === "string" ? v.uploadUrl : "";
  if (attachmentId <= 0 || !uploadUrl) return null;
  return { attachmentId, uploadUrl };
}

/** Human size, for the picker row and the limit message. */
export function formatSize(bytes: number): string {
  if (!Number.isFinite(bytes) || bytes <= 0) return "0 B";
  const units = ["B", "KB", "MB", "GB"];
  let n = bytes;
  let i = 0;
  while (n >= 1024 && i < units.length - 1) {
    n /= 1024;
    i += 1;
  }
  return `${i === 0 ? Math.round(n) : n.toFixed(1)} ${units[i]}`;
}

/** Whether an attachment is worth rendering inline in the chat bubble. */
export function isViewableImage(contentType: string): boolean {
  return contentType.startsWith("image/") && contentType !== "image/svg+xml";
}
